### ISSUE TITLE
Significant performance regression for datashader in numba 0.49.x

### ISSUE THREAD

**Issue by philippjfr** [numba/numba#5746]
Using latest released numba (0.49.1) I'm seeing a **significant** performance regression when compared to numba 0.48 in a simple datashader aggregation example. I'm downgrading from 0.49.1 to 0.48 with `conda install numba=0.48 --no-deps` to ensure no other packages change. The simple test case I'm using is the following:

```python
import timeit

from functools import partial

import datashader as ds
import numba
import numpy as np
import pandas as pd

canvas = ds.Canvas(plot_height=1000, plot_width=1000)

def agg(df):
    canvas.points(df, 'x', 'y', agg=ds.mean('value'))

def test_agg_performance(N, repeats=10):
    df = pd.DataFrame({'x': np.random.randn(N), 'y': np.random.randn(N), 'value': np.random.rand(N)})
    agg(df) # Warm up JIT
    return timeit.timeit(partial(agg, df), number=repeats)/repeats

print(f'Numba version: {numba.__version__}')
[(n, test_agg_performance(int(n))) for n in np.logspace(0, 8, 9)]
```

Here is a graph of the performance difference by the number of points being aggregated:

![bokeh_plot - 2020-05-25T153633 150](https://user-images.githubusercontent.com/1550771/82817591-895f4e00-9e9d-11ea-809a-b89e89f9e548.png)

And here is the notebook which I used to generate the plot: https://anaconda.org/philippjfr/profiling_numba/notebook

**Comment by stuartarchibald** [comment]
Thanks for the report. Which threading backend are you using?

**Comment by esc** [comment]
Also, which Datashader version can this be reproduced with?

**Comment by philippjfr** [comment]
>Thanks for the report. Which threading backend are you using?

How do I check? I looked at `numba.config.THREADING_LAYER` which is just `'default'`.

>Also, which Datashader version can this be reproduced with?

I'm using datashader master, but I just checked 0.10.0 and the results are the same.

**Comment by stuartarchibald** [comment]
Run this *after* you've executed some Numba code.
```
from numba import threading_layer
print("Threading layer chosen: %s" % threading_layer())
```

**Comment by stuartarchibald** [comment]
I'm pretty sure I know what the cause will be (TBB and performing nested parallelism as requested from user code), but am going to check.

**Comment by philippjfr** [comment]
Apparently no threading layer, even after executing the aggregation I get:

```
~/miniconda3/envs/rapidsai13/lib/python3.6/site-packages/numba/np/ufunc/parallel.py in threading_layer()
    322     """
    323     if _threading_layer is None:
--> 324         raise ValueError("Threading layer is not initialized.")
    325     else:
    326         return _threading_layer

ValueError: Threading layer is not initialized.
```

**Comment by philippjfr** [comment]
Actually yeah, I don't think we use parallelism for points aggregation or rather the only parallelism we use there is dask partitioning.

**Comment by stuartarchibald** [comment]
ok, thanks, guess it's not what I thought it was, which is good news as that's a very hard thing to fix :)

**Comment by esc** [comment]
FWIW I'm bisecting now.

**Comment by stuartarchibald** [comment]
Will leave it with you @esc, just wanted to make sure it wasn't threading!

Potential things off the top of my head:
 * SSA
 * Ref counting (think that invalid pass got removed in 0.49, perhaps DS was relying on it?)

**Comment by sklam** [comment]
> Ref counting (think that invalid pass got removed in 0.49, perhaps DS was relying on it?)

i think that is the most likely case.

**Comment by esc** [comment]
So my `git bisect` run led to the following commit:

https://github.com/numba/numba/commit/8c74571a2f97cede08c491b6c018ed144072e8bd

**Comment by esc** [comment]
Which in turn led to the following PR:

https://github.com/numba/numba/pull/5338

**Comment by esc** [comment]
Yeah, reverting 8c74571a2f97cede08c491b6c018ed144072e8bd on master get the performance back.

**Comment by stuartarchibald** [comment]
Fixed in #5795. Will make it into 0.50.0.
